### ISSUE TITLE
build error mexopencv.make opencv_aruco310

### ISSUE THREAD

**Issue by sushant4788** [kyamagu/mexopencv#293]
I followed the instructions from the wiki page to build mexopencv. I am getting the following error message while using `mexopencv.make`

OpenCV : github [https://github.com/opencv/opencv]
mexopencv : github [https://github.com/kyamagu/mexopencv] 

I need help in debugging this error! 
Thanks!

``` matlab
mexopencv.make('opencv_path','C:\git_opencv31\manual_build\install', 'opencv_contrib',false)
```
```
mex -O -largeArrayDims   -I'C:\mexopencv\include' -I'C:\git_opencv31\manual_build\install\include' -L'C:\git_opencv31\manual_build\install\x64\vc14\lib'  -lopencv_aruco310 -lopencv_bgsegm310 -lopencv_bioinspired310 -lopencv_calib3d310 -lopencv_ccalib310 -lopencv_core310 -lopencv_datasets310 -lopencv_dnn310 -lopencv_dpm310 -lopencv_face310 -lopencv_features2d310 -lopencv_flann310 -lopencv_fuzzy310 -lopencv_highgui310 -lopencv_imgcodecs310 -lopencv_imgproc310 -lopencv_line_descriptor310 -lopencv_ml310 -lopencv_objdetect310 -lopencv_optflow310 -lopencv_phase_unwrapping310 -lopencv_photo310 -lopencv_plot310 -lopencv_reg310 -lopencv_rgbd310 -lopencv_saliency310 -lopencv_shape310 -lopencv_stereo310 -lopencv_stitching310 -lopencv_structured_light310 -lopencv_superres310 -lopencv_surface_matching310 -lopencv_text310 -lopencv_tracking310 -lopencv_video310 -lopencv_videoio310 -lopencv_videostab310 -lopencv_xfeatures2d310 -lopencv_ximgproc310 -lopencv_xobjdetect310 -lopencv_xphoto310  -c 'C:\mexopencv\src\MxArray.cpp' -outdir 'C:\mexopencv\lib'
Error using mex
MEX cannot find library 'opencv_aruco310' specified with the -l option.
 MEX looks for a file with one of the names:
 opencv_aruco310.lib
 libopencv_aruco310.lib
 Please specify the path to this library with the -L option.


Error in mexopencv.make (line 96)
            if ~opts.dryrun, eval(cmd); end 
```

**Comment by amroamroamro** [comment]
I would help to see what files you have built/installed.

Please run the following commands in MATLAB, and post the output here:

``` matlab
>> !tree C:\git_opencv31\manual_build\install
>> !dir C:\git_opencv31\manual_build\install\x64\vc14\bin
>> !dir C:\git_opencv31\manual_build\install\x64\vc14\lib
```

**Comment by sushant4788** [comment]
@amroamroamro I ran the commands. The output is posted below. Thanks !! 

>> !tree C:\git_opencv31\manual_build\install
Folder PATH listing for volume Windows8_OS 
Volume serial number is C096-4CFD 
C:\GIT_OPENCV31\MANUAL_BUILD\INSTALL 
├───bin 
├───etc 
│   ├───haarcascades 
│   └───lbpcascades 
├───include 
│   ├───opencv 
│   └───opencv2 
│       ├───aruco 
│       ├───bioinspired 
│       ├───calib3d 
│       ├───ccalib 
│       ├───core 
│       │   ├───cuda 
│       │   │   └───detail 
│       │   └───hal 
│       ├───datasets 
│       ├───dnn 
│       ├───face 
│       ├───features2d 
│       ├───flann 
│       ├───fuzzy 
│       ├───highgui 
│       ├───imgcodecs 
│       ├───imgproc 
│       │   ├───detail 
│       │   └───hal 
│       ├───line_descriptor 
│       ├───ml 
│       ├───objdetect 
│       ├───optflow 
│       ├───phase_unwrapping 
│       ├───photo 
│       ├───reg 
│       ├───rgbd 
│       ├───saliency 
│       ├───shape 
│       ├───stereo 
│       ├───stitching 
│       │   └───detail 
│       ├───structured_light 
│       ├───superres 
│       ├───surface_matching 
│       ├───text 
│       ├───tracking 
│       ├───video 
│       ├───videoio 
│       ├───videostab 
│       ├───xfeatures2d 
│       ├───ximgproc 
│       └───xphoto 
└───x64 
    └───vc14 
        ├───bin 
        └───lib 

>> !dir C:\git_opencv31\manual_build\install\x64\vc14\bin
 Volume in drive C is Windows8_OS 
 Volume Serial Number is C096-4CFD 
 
 Directory of C:\git_opencv31\manual_build\install\x64\vc14\bin 
 
11/10/2016  02:45 PM    <DIR>          . 
11/10/2016  02:45 PM    <DIR>          .. 
11/10/2016  02:42 PM         1,429,504 opencv_aruco310d.dll 
11/10/2016  02:42 PM           507,904 opencv_bgsegm310d.dll 
11/10/2016  02:42 PM           610,304 opencv_bioinspired310d.dll 
11/10/2016  02:42 PM         4,250,624 opencv_calib3d310d.dll 
11/10/2016  02:42 PM         1,799,168 opencv_ccalib310d.dll 
11/10/2016  02:41 PM        16,388,608 opencv_core310d.dll 
11/10/2016  02:42 PM         1,707,520 opencv_datasets310d.dll 
11/10/2016  02:41 PM         6,750,720 opencv_dnn310d.dll 
11/10/2016  02:42 PM           771,072 opencv_dpm310d.dll 
11/10/2016  02:42 PM           908,800 opencv_face310d.dll 
11/10/2016  02:42 PM         2,374,656 opencv_features2d310d.dll 
11/10/2016  02:35 PM        16,318,976 opencv_ffmpeg310_64.dll 
11/10/2016  02:41 PM         1,740,288 opencv_flann310d.dll 
11/10/2016  02:41 PM           467,456 opencv_fuzzy310d.dll 
11/10/2016  02:42 PM           437,248 opencv_highgui310d.dll 
11/10/2016  02:41 PM         5,735,424 opencv_imgcodecs310d.dll 
11/10/2016  02:41 PM        31,412,224 opencv_imgproc310d.dll 
11/10/2016  02:42 PM         1,291,776 opencv_line_descriptor310d.dll 
11/10/2016  02:41 PM         1,863,168 opencv_ml310d.dll 
11/10/2016  02:42 PM         1,703,424 opencv_objdetect310d.dll 
11/10/2016  02:42 PM         1,475,584 opencv_optflow310d.dll 
11/10/2016  02:42 PM           551,936 opencv_phase_unwrapping310d.dll 
11/10/2016  02:41 PM         1,996,800 opencv_photo310d.dll 
11/10/2016  02:42 PM           440,320 opencv_plot310d.dll 
11/10/2016  02:41 PM           651,264 opencv_reg310d.dll 
11/10/2016  02:42 PM         2,015,744 opencv_rgbd310d.dll 
11/10/2016  02:42 PM         1,023,488 opencv_saliency310d.dll 
11/10/2016  02:41 PM           926,720 opencv_shape310d.dll 
11/10/2016  02:42 PM           769,536 opencv_stereo310d.dll 
11/10/2016  02:42 PM         2,835,456 opencv_stitching310d.dll 
11/10/2016  02:43 PM           800,768 opencv_structured_light310d.dll 
11/10/2016  02:42 PM           704,512 opencv_superres310d.dll 
11/10/2016  02:41 PM         1,485,312 opencv_surface_matching310d.dll 
11/10/2016  02:42 PM         1,978,880 opencv_text310d.dll 
11/10/2016  02:43 PM         5,676,544 opencv_tracking310d.dll 
11/10/2016  02:41 PM           952,832 opencv_video310d.dll 
11/10/2016  02:42 PM           807,424 opencv_videoio310d.dll 
11/10/2016  02:42 PM         1,617,920 opencv_videostab310d.dll 
11/10/2016  02:42 PM         4,316,672 opencv_xfeatures2d310d.dll 
11/10/2016  02:42 PM         3,018,752 opencv_ximgproc310d.dll 
11/10/2016  02:42 PM           792,064 opencv_xobjdetect310d.dll 
11/10/2016  02:42 PM         1,865,216 opencv_xphoto310d.dll 
              42 File(s)    135,172,608 bytes 
               2 Dir(s)  154,670,923,776 bytes free 

>> !dir C:\git_opencv31\manual_build\install\x64\vc14\lib
 Volume in drive C is Windows8_OS 
 Volume Serial Number is C096-4CFD 
 
 Directory of C:\git_opencv31\manual_build\install\x64\vc14\lib 
 
11/10/2016  02:45 PM    <DIR>          . 
11/10/2016  02:45 PM    <DIR>          .. 
11/10/2016  02:36 PM               433 OpenCVConfig-version.cmake 
11/10/2016  02:38 PM             9,519 OpenCVConfig.cmake 
11/10/2016  02:39 PM            29,926 OpenCVModules-debug.cmake 
11/10/2016  02:39 PM             7,334 OpenCVModules.cmake 
11/10/2016  02:42 PM           269,594 opencv_aruco310d.lib 
11/10/2016  02:42 PM           203,814 opencv_bgsegm310d.lib 
11/10/2016  02:42 PM           180,564 opencv_bioinspired310d.lib 
11/10/2016  02:42 PM           298,208 opencv_calib3d310d.lib 
11/10/2016  02:42 PM           286,270 opencv_ccalib310d.lib 
11/10/2016  02:41 PM           661,938 opencv_core310d.lib 
11/10/2016  02:42 PM           244,166 opencv_datasets310d.lib 
11/10/2016  02:41 PM         1,588,594 opencv_dnn310d.lib 
11/10/2016  02:42 PM           183,208 opencv_dpm310d.lib 
11/10/2016  02:42 PM           198,796 opencv_face310d.lib 
11/10/2016  02:42 PM           329,092 opencv_features2d310d.lib 
11/10/2016  02:41 PM           189,362 opencv_flann310d.lib 
11/10/2016  02:41 PM           180,256 opencv_fuzzy310d.lib 
11/10/2016  02:42 PM           200,594 opencv_highgui310d.lib 
11/10/2016  02:41 PM           186,356 opencv_imgcodecs310d.lib 
11/10/2016  02:41 PM           312,310 opencv_imgproc310d.lib 
11/10/2016  02:42 PM           286,574 opencv_line_descriptor310d.lib 
11/10/2016  02:41 PM           223,462 opencv_ml310d.lib 
11/10/2016  02:42 PM           266,604 opencv_objdetect310d.lib 
11/10/2016  02:42 PM           285,510 opencv_optflow310d.lib 
11/10/2016  02:42 PM           201,040 opencv_phase_unwrapping310d.lib 
11/10/2016  02:41 PM           260,636 opencv_photo310d.lib 
11/10/2016  02:42 PM           179,076 opencv_plot310d.lib 
11/10/2016  02:41 PM           206,076 opencv_reg310d.lib 
11/10/2016  02:42 PM           315,678 opencv_rgbd310d.lib 
11/10/2016  02:42 PM           237,456 opencv_saliency310d.lib 
11/10/2016  02:41 PM           221,456 opencv_shape310d.lib 
11/10/2016  02:42 PM           237,468 opencv_stereo310d.lib 
11/10/2016  02:42 PM           759,806 opencv_stitching310d.lib 
11/10/2016  02:43 PM           289,436 opencv_structured_light310d.lib 
11/10/2016  02:42 PM           270,074 opencv_superres310d.lib 
11/10/2016  02:41 PM           206,220 opencv_surface_matching310d.lib 
11/10/2016  02:42 PM           284,782 opencv_text310d.lib 
11/10/2016  02:43 PM           349,954 opencv_tracking310d.lib 
11/10/2016  02:41 PM           215,668 opencv_video310d.lib 
11/10/2016  02:42 PM           194,850 opencv_videoio310d.lib 
11/10/2016  02:42 PM           521,552 opencv_videostab310d.lib 
11/10/2016  02:42 PM           329,198 opencv_xfeatures2d310d.lib 
11/10/2016  02:42 PM           338,690 opencv_ximgproc310d.lib 
11/10/2016  02:42 PM           195,706 opencv_xobjdetect310d.lib 
11/10/2016  02:42 PM           192,872 opencv_xphoto310d.lib 
              45 File(s)     12,630,178 bytes 
               2 Dir(s)  154,670,788,608 bytes free

**Comment by amroamroamro** [comment]
A couple of notes:

- You must build against OpenCV **3.1.0**, not the `master` dev version. I suspect that because you have a `phase_unwrapping` module which shouldn't be there.

- it looks like you only compiled the "Debug" version (the "d" in `opencv_*310d`).  If you wanted to use the debug version, you must call `mexopencv.make(.., 'Debug',true)`. Otherwise I suggest you switch to the "Release" version as it's optimized and faster.

- it seems you compiled both opencv and opencv_contrib modules, so you should set `mexopencv.make(.., 'opencv_contrib',true)` not false.

**Comment by sushant4788** [comment]
@amroamroamro
Hi,
As you suggested, I tried to build against OpenCV3.1.0 by downloading it from http://opencv.org/downloads.html
While building the opencv source I got the following error. 
![capture](https://cloud.githubusercontent.com/assets/5024836/20222604/f3690514-a803-11e6-97e4-f32c7d77fcd2.PNG)
I believe I saw this error earlier, and thought of building against the master. 
Please let me know if I am missing something. 
Thanks

**Comment by amroamroamro** [comment]
@sushant4788 
The 3.1.0 version requirement applies to both opencv and opencv_contrib (simultaneously). This is explained in the "Download OpenCV" section in [this wiki page](https://github.com/kyamagu/mexopencv/wiki/Installation-%28Windows%2C-MATLAB%2C-OpenCV-3%29#download-opencv).

The error you saw is because you attempt to compile opencv-3.1.0 with opencv_contrib-master!

Just grab the "source code" (ZIP files) of the 3.1.0 releases from here:
- https://github.com/opencv/opencv/releases/tag/3.1.0
- https://github.com/opencv/opencv_contrib/releases/tag/3.1.0

Bottom line is you can't build mexopencv against the master branch of opencv/opencv_contrib, only stable releases.

**Comment by sushant4788** [comment]
@amroamroamro 
Thanks a lot. The issue is resolved. It was the mismatch of the branches for opencv/opencv_contrib. Thanks again.

**Comment by amroamroamro** [comment]
Great, closing as solved.
